client/src/contexts/UndoRedoContext.tsx

```diff
@@ -1,103 +1,249 @@
 
+  // these states are used to memorize the content of recently modified text
+  // so that we can use this old value to prepare a propper ModifiedCommand
+  const [observedTextId, setObservedTextId] = useState<string | undefined>(undefined);
+  const [observedTextContent, setObservedTextContent] = useState<string | undefined>(undefined);
+
+  const [recentlyActiveObjects, setRecentlyActiveObjects] = useState<any[]>([]);
+
   const canvasRef = useRef<fabric.Canvas | null>(canvas);
-  const undoStack = useRef<Array<any>>([]);
-  const redoStack = useRef<Array<any>>([]);
+  const undoStack = useRef<Array<UndoRedoCommand>>([]);
+  const redoStack = useRef<Array<UndoRedoCommand>>([]);
 
   useEffect(() => {
     canvasRef.current = canvas;
   }, [canvas]);
 
-  const areStatesEqual = (state1: any, state2: any) => {
-    return JSON.stringify(state1) === JSON.stringify(state2);
-  };
-
-  const saveState = useCallback(() => {
-    const canvas = canvasRef.current;
-    if (!canvas) {
-      return;
-    }
-    const state = canvas.toJSON();
-    if (undoStack.current.length === 0 || !areStatesEqual(state, undoStack.current[undoStack.current.length - 1])) {
-      undoStack.current.push(state);
-      if (undoStack.current.length > 50) {
-        undoStack.current.shift();
-      }
-      redoStack.current.length = 0;
+  const saveCommand = useCallback((command: UndoRedoCommand) => {
+    undoStack.current.push(command);
+    if (undoStack.current.length > 50) {
+      undoStack.current.shift();
     }
+    redoStack.current.length = 0;
   }, []);
 
   const undo = useCallback(() => {
     const canvas = canvasRef.current;
-    if (canvas && undoStack.current.length > 0) {
-      const currentState = canvas.toJSON();
-      redoStack.current.push(currentState);
 
-      const state = undoStack.current.pop();
-      if (!state) {
-        return;
-      }
-      canvas.loadFromJSON(state, () => {
-        canvas.renderAll();
-        canvas.calcOffset();
-      });
+    if (!canvas || undoStack.current.length === 0) {
+      return;
     }
+    const lastAction = undoStack.current.pop();
+    if (!lastAction) {
+      return;
+    }
+
+    lastAction.undo();
+    redoStack.current.push(lastAction);
+    canvas.renderAll();
   }, []);
 
   const redo = useCallback(() => {
     const canvas = canvasRef.current;
-    if (canvas && redoStack.current.length > 0) {
-      const currentState = canvas.toJSON();
-      undoStack.current.push(currentState);
-
-      const state = redoStack.current.pop();
-      if (!state) {
-        return;
-      }
-      canvas.loadFromJSON(state, () => {
-        canvas.renderAll();
-        canvas.calcOffset();
-      });
+    if (!canvas || redoStack.current.length === 0) {
+      return;
+    }
+    const lastAction = redoStack.current.pop();
+    if (!lastAction) {
+      return;
     }
+
+    lastAction.redo();
+    undoStack.current.push(lastAction);
+    canvas.renderAll();
   }, []);
 
   useEffect(() => {
```

Comment:
There’s quite a bit of complex code in this file.  Maybe we could create a custom hook for this useEffect? The logic inside isn’t shared with other components or contexts, and it’s focused on managing listeners.